### PR TITLE
Implementing CN timeout bonifications

### LINKED ISSUE #664: CN Timeout bonifications
**Proposal:**
Increase timeout with a fixed static value (based on blocktime) for each successful message received.

**Motivation:**
Avoid timeout when the CNs are running on normal operation.

**Negative impacts**
A minor negative impact on `change view` timeout synchronization.

However, it will not affect liveness due to the nature of the agreement using the exponential formula, which will not be affected by fixed increments.

**Details**

* When a node receives `OnPrepRequest` it will increase its timeout by, for example, `ceil(blockTime/10 * 1)`;
* When a node receives a `OnPrepResponse` it will increase its timeout by, for example, `ceil(blockTime/10*2)`;
* When a node receives a `OnCommit` it will increase its timeout by, for example, `ceil(blockTime/10*3)`;

This will ensure that nodes are going to delay a little bit more to timeout when CNs are "supposedly" running normally.

What do you think, @erikzhang, @igormcoelho, @shargon, @jsolman?

### PR DESCRIPTION
**Pull request by vncoelho**
Closes #664

### CODE REVIEW

**Comment by vncoelho** [comment]
image

**Comment by erikzhang** [comment]
Why do we change timer everywhere?

**Comment by vncoelho** [comment]
The looked like to be the simplest way to implement.
Do you recommend another way, @erikzhang?

This change is important for longterm and the mathematic behind is strong.
It will not affect liveness.

**Comment by igormcoelho** [comment]
> Why do we change timer everywhere?

@erikzhang , we can rename it to IncreaseTimer, perhaps it's a better name.

**Comment by erikzhang** [comment]
Can anyone review this PR? However, I am worried that if we continue to optimize the consensus, dBFT2.0 may never be able to go to the mainnet.

**Comment by igormcoelho** [comment]
@erikzhang this is not just optimization... we designed this in order to prevent common change views, that we thought could help avoiding the last stall on test net.
But now, after reviewing the logs, we found an actual fix, on #706 / #707 

My suggestion is to ask NGD for a report over this... if network behaves so much better, we advise putting this on TestNet together with the pending fix. If it's not deemed too great, then it can be kept for later. It looks like an arbitrary change, but we thought about that from a long time, this was the only way to make our priv net work "perfectly".

**Comment by igormcoelho** [comment]
@jsolman @shargon can you check this, if you agree on the logic?

**Comment by vncoelho** [comment]
These guys have no limits, @erikzhang...aheuaheuaea
It is going to be an endless journey of optimization :D Let's make science !! :dancer: 
I am joking, @erikzhang....aehauheuaea

This is a nice change. The performance improved considerably as @igormcoelho highlighted.
Several unnecessary `changeview` are being avoided with this.
Let's move is forward as soon as possible. There are no negative impacts for real-world operation and it also represents a good step towards scalability.

**Comment by Qiao-Jin** [comment]
Excuse me, but will this have negative effect upon onchain TPS (Transactions per second)? Thx!

**Comment by jsolman** [comment]
No, in general it would probably improve tps.

File: neo/Consensus/ConsensusService.cs [hunk 1]
```diff
@@ -249,6 +265,14 @@ private void OnCommitReceived(ConsensusPayload payload, Commit commit)
             existingCommitPayload = payload;
         }
 
+        // this function increases existing timer (never decreases) with a value proportional to `maxDelayInBlockTimes`*`Blockchain.SecondsPerBlock`
+        private void IncreaseTimerWithBonification(int maxDelayInBlockTimes)
```
Comment (jsolman) [inline neo/Consensus/ConsensusService.cs]: `ExtendTimerByFactor`

Comment (vncoelho) [reply neo/Consensus/ConsensusService.cs]: Refactored, Jeff, thanks.

File: neo/Consensus/ConsensusService.cs [hunk 2]
```diff
@@ -227,6 +239,10 @@ private void OnCommitReceived(ConsensusPayload payload, Commit commit)
                 return;
             }
 
+            // Timeout bonification: commit has been received with sucess
```
[the remark this answers is not in the document]
Comment (vncoelho) [reply neo/Consensus/ConsensusService.cs]: Done, thanks.

File: neo/Consensus/ConsensusService.cs [hunk 3]
```diff
@@ -75,6 +81,10 @@ private bool AddTransaction(Transaction tx, bool verify)
                     // previously sent prepare request, then we don't want to send a prepare response.
                     if (context.IsPrimary() || context.WatchOnly()) return true;
 
+                    // Timeout bonification: prepare response has been sent with sucess
```
[the remark this answers is not in the document]
Comment (vncoelho) [reply neo/Consensus/ConsensusService.cs]: Done!

File: neo/Consensus/ConsensusService.cs [hunk 4]
```diff
@@ -75,6 +81,10 @@ private bool AddTransaction(Transaction tx, bool verify)
...
+                    // around 2*15/M=30.0/5 ~ 40% block time (for M=5)
+                    IncreaseTimerWithBonification(2);
```
[the remark this answers is not in the document]
Comment (vncoelho) [reply neo/Consensus/ConsensusService.cs]: Eahueahueaea, PortuEnglish, @jsolman.
Refactor has been made. Thanks.

**Comment by vncoelho** [comment]
Travis problems, how to re-run Travis? Sometime ago I saw that button on their system.

**Comment by shargon** [comment]
Restarted

**Comment by vncoelho** [comment]
I just noticed I was not logged in there, @shargon. Thanks,

**Comment by shargon** [comment]
travis is not working on neo-vm too...

**Comment by vncoelho** [comment]
Thanks everybody.
Byzantine is byzantine, agreement is agreement, fly we go.

**Pull request merged**